### DISCUSSION TITLE
pyrfc - call SE38 programs and take screenshots?

### DISCUSSION THREAD

**Discussion by chandra-a** [SAP/PyRFC#292] [Q&A] [answered]
hi All, 

a) would it be possible via pyrfc to also call se38 programs (if yes what would be the syntax), Or we can only call function modules?
b) Also,  after executing a FM, is there an option to take a 'screenshot' of the result? from what i understood the pyrfc FM call gives you the export value of FM but there are some UI based FMs that give a screenshot based result which i want to capture (say into a word doc)

appreciate any inutes

**Comment by bsrdjan** [comment]
Hello @chandra-a,

a) se38 programs (ABAP reports) can't be called via SAP RFC protocol, only function modules. To call ABAP report via RFC and PyRFC, an RFC enabled function module shall be created, that calls the ABAP report internally. Here one example: https://answers.sap.com/questions/5421837/how-to-call-a-report-through-rfc.html
b) I am not aware of ABAP Function Modules giving a screenshot based result. Do you have an example?

**Reply by chandra-a** [reply]
> Hello @chandra-a,
> 
> a) se38 programs (ABAP reports) can't be called via SAP RFC protocol, only function modules. To call ABAP report via RFC and PyRFC, an RFC enabled function module shall be created, that calls the ABAP report internally. Here one example: https://answers.sap.com/questions/5421837/how-to-call-a-report-through-rfc.html b) I am not aware of ABAP Function Modules giving a screenshot based result. Do you have an example?

Hi, thanks for the answer. TMS_UI_CHECK_TRANSPORT_TOOLS is one example where we get a UI response when the FM is executed.

Thanks

**Comment by Ulrich-Schmidt** [comment, marked as answer]
Hello,
"screenshot based result" is probably not the correct description for this, but there are indeed some old FMs/BAPIs which try to send Dynpro data (data formatted in DIAG protocol) back to the caller. When the caller is connected to the SAP System via RFC protocol (like in this case), this usually results in a dump DYNPRO_SEND_IN_BACKGROUND "Screen output without connection to user".

If you are seeing such dumps in ST22, then the following parameter for the RFC connection might help you:
USE_SAPGUI=1/2 (see sapnwrfc.ini file for details)

Basically there are two cases:

1. The screen output is not relevant and can be ignored. In that case set USE_SAPGUI=2. This starts a hidden SAPGui, which just receives the DIAG data and ignores it.
2. The screen output contains important response data. In that case set USE_SAPGUI=1 to have a visible SAPGui attached to the user session. Note however, that SAPGui is started as a child process of the Python process, so you will not have access to the DIAG data from within your Python program.  An end-user will have to sit in front of the screen and read the data from the Gui.

In any case: SAPGui needs to be installed on the machine, where your Python program is running. On Windows, you can use the standard SAPGui, on Linux/Unix the SAP Java Gui (Platin Gui).
